redis/src/main/scala/zio/redis/TestExecutor.scala

```diff
@@ -650,6 +660,193 @@ private[redis] final class TestExecutor private (
+              val key   = by.replace("*", next)
+              val value = strings.get(key)
+              aggr.flatMap(c => value.map(n => n.fold(c)(c :+ (next, _))))
+            }
+            pairs.map(_.sortBy(_._2).map(_._1))
+          }
+
+        def get(list: Chunk[String]) =
+          list.foldLeft(STM.succeedNow(Chunk[String]())) { case (aggr, next) =>
+            val keys = getPatterns.map(p => p.replace("*", next))
+            val values = keys.foldLeft(STM.succeedNow(Chunk[String]())) { case (all, k) =>
+              if (k == "#") all.map(a => a :+ next)
+              else all.flatMap(a => strings.get(k).map(v => v.fold(a)(a :+ _)))
+            }
+            aggr.flatMap(c => values.map(vs => c ++ vs))
+          }
+
+        def handle(listOpt: Option[Chunk[String]]) =
+          for {
+            list   <- listOpt.fold[STM[RespValue.Error, Chunk[String]]](STM.fail(Replies.Error))(STM.succeedNow)
+            sorted <- sort(list)
+            sliced  = limit.fold(sorted) { case (offset, count) => sorted.drop(offset).take(count) }
+            result <- get(sliced)
+          } yield result
+
+        val zstm = for {
+          keyInfoOpt <- keys.get(key)
+          keyInfo    <- keyInfoOpt.fold[STM[RespValue.Error, KeyInfo]](STM.fail(Replies.Error))(STM.succeedNow)
+          result <- keyInfo.`type` match {
+                      case KeyType.Lists => lists.get(key).flatMap(handle)
+                      case KeyType.Sets  => sets.get(key).flatMap(s => handle(s.map(Chunk.fromIterable)))
+                      case KeyType.SortedSets =>
+                        sortedSets.get(key).flatMap(s => handle(s.map(ks => Chunk.fromIterable(ks.keys))))
+                      case _ => STM.fail(Replies.Error)
+                    }
+          _ <- storeKey.fold(ZSTM.unit)(sk => putList(sk, result))
+        } yield if (storeKey.isEmpty) Replies.array(result) else RespValue.Integer(result.size.toLong)
+        zstm.catchAll(error => STM.succeedNow(error))
+
+      case api.Keys.Ttl =>
+        val key = input.head.asString
+        STM
+          .ifM(keys.contains(key))(
+            ttlOf(key, now).map(_.fold(-1L)(_.toSeconds)),
+            STM.succeedNow(-2L)
+          )
+          .map(RespValue.Integer)
+
+      case api.Keys.PTtl =>
+        val key = input.head.asString
+        STM
+          .ifM(keys.contains(key))(
+            ttlOf(key, now).map(_.fold(-1L)(_.toMillis)),
+            STM.succeedNow(-2L)
+          )
+          .map(RespValue.Integer)
+
+      case api.Keys.Persist =>
+        val key = input.head.asString
+        keys
+          .get(key)
+          .flatMap(_.fold(STM.succeedNow(0L))(info => keys.put(key, info.copy(expireAt = None)).as(1L)))
+          .map(RespValue.Integer)
+
+      case api.Keys.Expire =>
+        val key      = input(0).asString
+        val unixtime = now.plusSeconds(input(1).asLong)
```

Comment:
This should be suspended, but since we don't have a proper solution for clock in stm, it can stay.